### PR TITLE
Start integrating QArray into utils 

### PR DESCRIPTION
**Pull request by pierreguilmin**
Replaces #621. Documentation will come later (see incoming PR).

I ported
- [x] `utils/utils/general.py`
- [x] `_utils.py`
- [x] `_checks.py`
- [x] `utils/vectorization.py`
- [x] `utils/jax_utils.py`

Remaining things from these files (but I'm tired and this PR is already massive, tasks added to https://linear.app/dynamiqs/issue/DYN-210/qarray-object):
- [ ] `ptrace()` in `utils/utils/general.py` and in `QArray` API
- [ ] `expect()` vmap is bugged

I also added a few technical points to https://linear.app/dynamiqs/issue/DYN-210/qarray-object.

Few random remarks:
- Note that for now vectorization depends on `eye(n)` returning a `QArray` which will soon be fixed.
- I had to be a bit siouks to avoid circular imports, any other solution is welcome. :)
- Doctest can be quickly run for `utils/utils/general.py` by implementing `DenseQArray.__repr__` with `repr(self.data)`, which I did in this PR for convenience. This will be fixed soon.

For later PRs (tasks added to https://linear.app/dynamiqs/issue/DYN-210/qarray-object):
- [ ] implement a proper `__repr__` and fix the docstrings
- [ ]  creation utilities: `utils/operators.py`, `utils/states.py`, `utils/optimal_control.py`, `dark/dark.py`
- [ ] `utils/wigner.py`
- [ ] `plots/`
- [ ] `utils/random.py`
- [ ] `sesolve/`, `mesolve/`, `smesolve/`, `core/`, `result.py`, `time_array.py`
- [ ] sparse DIA
- [ ] make objects in `types.py` public and documented
- [ ] fix the tutorials and the documentation everywhere

### CODE REVIEW

File: dynamiqs/qarrays/types.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    # if not (isbra(data) or isket(data) or isdm(data) or isop(data)):
+    # raise ValueError(
+    #     f'DenseQArray data must be a bra, a ket, a density matrix '
+    #     f'or and operator. Got array with size {data.shape}'
+    # )
+    # TODO: check that dims and shape work
+
+    x = jnp.asarray(x)
+    if dims is None:
+        dims = (x.shape[-2],)
```
Comment (gautierronan) [inline dynamiqs/qarrays/types.py:58]: This would not be correct if this input is a bra of shape (1, n).

**Review by gautierronan — CHANGES_REQUESTED** [review]
Looks good to me! Some minor comments below

File: dynamiqs/qarrays/types.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+_NestedQArrayLikeList = list[Union[_QArrayLike, '_NestedQArrayLikeList']]
+# a type alias for any type compatible with asqarray
+QArrayLike = Union[_QArrayLike, _NestedQArrayLikeList]
+
+
+def is_qarraylike(x: Any) -> bool:
+    if isinstance(x, get_args(_QArrayLike)):
+        return True
+    elif isinstance(x, list):
+        return all(isinstance(_x, get_args(_QArrayLike)) for _x in x)
```
[the remark this answers is not in the document]
Comment (pierreguilmin) [reply dynamiqs/qarrays/types.py:37]: Your fix does account for nested list no?

File: dynamiqs/qarrays/types.py (hunk 1 above)
Comment (pierreguilmin) [reply dynamiqs/qarrays/types.py:58]: True, let me add a todo fix to move on quickly.

**Pull request merged**